rasa/utils/tensorflow/layers.py

```diff
@@ -198,24 +188,149 @@ class DenseWithSparseWeights(tf.keras.layers.Dense):
+            bias_constraint: Constraint function applied to the bias vector.
+        """
         super().__init__(**kwargs)
-        self.sparsity = sparsity
+
+        if density < 0.0 or density > 1.0:
+            raise TFLayerConfigException("Layer density must be in [0, 1].")
+
+        self.density = density
 
     def build(self, input_shape: tf.TensorShape) -> None:
+        """Prepares the kernel mask.
+
+        Args:
+            input_shape: Shape of the inputs to this layer
+        """
         super().build(input_shape)
-        # create random mask to set fraction of the `kernel` weights to zero
-        kernel_mask = tf.random.uniform(tf.shape(self.kernel), 0, 1)
-        kernel_mask = tf.cast(
-            tf.greater_equal(kernel_mask, self.sparsity), self.kernel.dtype
+
+        if self.density == 1.0:
+            self.kernel_mask = None
+            return
+
+        kernel_shape = tf.shape(self.kernel)
+        num_rows = kernel_shape[0].numpy()
+        num_cols = kernel_shape[1].numpy()
+
+        # Construct mask with given density and guarantee that every output is
+        # connected to at least one input
+        num_connected_per_output = max(
+            1, tf.cast(tf.math.ceil(self.density * num_rows), tf.int32)
         )
+        kernel_mask = self._random_mask_with_no_empty_rows(
+            num_rows, num_cols, num_connected_per_output
+        )
+
+        # Guarantee that every input is also connected to at least one output
+        kernel_mask += self._missing_connections(num_rows, num_cols, kernel_mask)
+
         self.kernel_mask = tf.Variable(
             initial_value=kernel_mask, trainable=False, name="kernel_mask"
         )
 
+    def _random_mask_with_no_empty_rows(
+        self, num_rows: int, num_cols: int, num_connected_per_output: int
+    ) -> tf.Tensor:
+        """Creates a random mask with given number of connections per output.
+
+        Args:
+            num_rows: Number of rows in the mask
+            num_cols: Number of columns in the mask
+            num_connected_per_output: Number of ones in each column
+
+        Returns:
+            A random mask matrix
+        """
+        num_disconnected_per_output = num_rows - num_connected_per_output
+
+        # We create each column separately
+        # as [1 1 1 ... 1 0 0 0 ... 0] and then shuffle it randomly
+        kernel_mask_cols = [
+            tf.random.shuffle(
+                tf.concat(
+                    [
+                        tf.ones(
+                            shape=num_connected_per_output, dtype=self.kernel.dtype
+                        ),
+                        tf.zeros(
+                            shape=num_disconnected_per_output, dtype=self.kernel.dtype
+                        ),
+                    ],
+                    axis=0,
+                )
+            )
+            for _ in range(num_cols)
+        ]
+        return tf.transpose(tf.stack(kernel_mask_cols, axis=0))
```

Comment:
then you should be able to concat over `axis=1`
```suggestion
        return tf.concat(kernel_mask_cols, axis=1)
```